feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2021 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package notify
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/labstack/echo/v4"
+
+	"magma/feg/gateway/sbi"
+	n7_client "magma/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl"
+	"magma/feg/gateway/services/n7_n40_proxy/n7"
+	"magma/feg/gateway/services/session_proxy/relay"
+	"magma/gateway/service_registry"
+	"magma/lte/cloud/go/protos"
+)
+
+// notify_handler implements the following
+// - Creates a HTTP server to receive notifications from PCF
+// - Handles SmPolicyUpdateNotification from PCF convert the message to proto and sends it as
+//   PolicyReAuth to SessionProxyResponder(feg_relay)
+// - Handles SmPolicyTerminateNotification from PCF convert the message to prot and sends it as
+//   AbortSession to feg_relay.
+
+const (
+	NOTIFY_PATH      = "/"
+	UnableToDeliver  = 3002
+	EncodedSessionId = "encodedSessionId"
+	// Shutdown timeout is used as a timeout to exit the Echo Server Shutdown
+	ShutdownTimeout = 10 * time.Second
+)
+
+type NotificationHandler struct {
+	config        *n7.N7ClientConfig
+	notifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+}
+
+func NewNotificationHandler(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) *NotificationHandler {
+	return &NotificationHandler{
+		config:        config,
+		notifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+}
```

Comment:
I still think that NewNotificationHandler could encapsulate all that (if you want you can create a new function called NewStartedNotificationHandlerWithHanlders which will run NewNotificationHandler and add the handlers and start

Will you ever need a non registered, non started handler?

A hint for that is that you are repeating the same code in the test than in the controller. Remember, if you see repeated code, you probably are missing an opportunity to encapsulate an object